### PR TITLE
Fix NPE in KeyValue when the message payload is null

### PR DESCRIPTION
**Pull request by vroyer**
### Motivation

Fix a NullPointerException when decoding a KeyValue with a null message payload.

### Modifications

Return a KeyValue(null, null) when data is null or empty.

### Verifying this change

Add a unit test in org.apache.pulsar.common.schema.KeyValueTest

### CODE REVIEW

File: pulsar-client-api/src/main/java/org/apache/pulsar/common/schema/KeyValue.java [hunk 1]
```diff
@@ -33,6 +33,8 @@
     private final K key;
     private final V value;
 
+    public static final KeyValue NULL = new KeyValue(null, null);
```
Comment (eolivelli) [inline pulsar-client-api/src/main/java/org/apache/pulsar/common/schema/KeyValue.java:36]: can we make this `private` ?

File: pulsar-client-api/src/main/java/org/apache/pulsar/common/schema/KeyValue.java [hunk 2]
```diff
@@ -130,6 +132,10 @@ public String toString() {
      * @return the decoded key/value pair
      */
     public static <K, V> KeyValue<K, V> decode(byte[] data, KeyValueDecoder<K, V> decoder) {
+        if (data == null || data.length == 0) {
```
Comment (eolivelli) [inline pulsar-client-api/src/main/java/org/apache/pulsar/common/schema/KeyValue.java:135]: can you please explain how it can happen that we fall into these cases ?

I mean, probably data == null is possible in case of a message will NULL value, but I cannot think about a case in which the array may be of zero length.

probably in case of zero length array we should throw an IllegalArgumentException, because it is not possible that it is a valid value, as we are always writing the length of the key and the length of the value (with -1 as marker for nulls)

File: pulsar-client/src/test/java/org/apache/pulsar/common/schema/KeyValueTest.java [hunk 3]
```diff
@@ -53,6 +56,7 @@
 import org.apache.pulsar.client.impl.schema.StringSchema;
 import org.apache.pulsar.client.impl.schema.TimeSchema;
 import org.apache.pulsar.client.impl.schema.TimestampSchema;
+import org.checkerframework.checker.units.qual.K;
```
Comment (eolivelli) [inline pulsar-client/src/test/java/org/apache/pulsar/common/schema/KeyValueTest.java:59]: please remove this import

File: pulsar-client/src/test/java/org/apache/pulsar/common/schema/KeyValueTest.java [hunk 4]
```diff
@@ -29,6 +31,7 @@
 import java.time.LocalDate;
 import java.time.LocalDateTime;
 import java.time.LocalTime;
+import java.util.ArrayList;
```
Comment (eolivelli) [inline pulsar-client/src/test/java/org/apache/pulsar/common/schema/KeyValueTest.java:34]: this import looks unused

File: pulsar-client/src/test/java/org/apache/pulsar/common/schema/KeyValueTest.java [hunk 5]
```diff
@@ -20,7 +20,9 @@
 
 import static java.nio.charset.StandardCharsets.UTF_8;
 import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotNull;
 
+import com.google.common.collect.ImmutableList;
```
Comment (eolivelli) [inline pulsar-client/src/test/java/org/apache/pulsar/common/schema/KeyValueTest.java:25]: this import looks unused

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: pulsar-client-api/src/main/java/org/apache/pulsar/common/schema/KeyValue.java (hunk 2 above)
Comment (vroyer) [reply pulsar-client-api/src/main/java/org/apache/pulsar/common/schema/KeyValue.java:135]: It appends when key is SEPARATED and value is null, here is the KeyValueSchema.encode():

```
public byte[] encode(KeyValue<K, V> message) {
        if (keyValueEncodingType != null && keyValueEncodingType == KeyValueEncodingType.INLINE) {
            return KeyValue.encode(
                message.getKey(),
                keySchema,
                message.getValue(),
                valueSchema
            );
        } else {
            if (message.getValue() == null) {
                return null;
            }
            return valueSchema.encode(message.getValue());
        }
    }
```

Comment (eolivelli) [reply pulsar-client-api/src/main/java/org/apache/pulsar/common/schema/KeyValue.java:135]: @vroyer I still do not see how we can have a zero byte array from your example.

if we have `keyValueEncodingType == KeyValueEncodingType.SEPARATED` then we are not using the KeyValue encoding,  but the value contains only the serialized version of the value (`valueSchema.encode(message.getValue())` ).

this "decode" code should be used only when you read an KeyValueEncodingType.INLINE message, that contains both the key and the value inside the payload of the message:
- keylen
- key bytes
- valuelen
- value bytes

when you serialize a null KeyValue you can a null value
when you serialize a null key you see -1 as "keylen" 
when you serialize a null value you see -1 as "valuelen"

there is no way to see a zero byte array

Comment (vroyer) [reply pulsar-client-api/src/main/java/org/apache/pulsar/common/schema/KeyValue.java:135]: Yes, https://github.com/apache/pulsar/pull/10089 is possibly the root cause. The KeyValue with a null value was decoded as INLINED, so this protection https://github.com/apache/pulsar/blob/master/pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/KeyValueSchema.java#L146 was not working as expected.

**Comment by eolivelli** [comment]
it looks like we can close this PR in favour of #10089

correct  ?

File: pulsar-client/src/test/java/org/apache/pulsar/common/schema/KeyValueTest.java [hunk 6]
```diff
@@ -105,6 +106,18 @@ public void testAllSchemas(Map<Schema, List<Object>> schemas) {
         }
     }
 
+    @Test
+    public void testNullPayload() {
+        KeyValue kv = KeyValue.decode(
+                null,
```
Comment (sijie) [inline pulsar-client/src/test/java/org/apache/pulsar/common/schema/KeyValueTest.java:112]: Don't you need to test empty data array?

File: pulsar-client/src/test/java/org/apache/pulsar/common/schema/KeyValueTest.java [hunk 7]
```diff
@@ -105,6 +106,18 @@ public void testAllSchemas(Map<Schema, List<Object>> schemas) {
...
+                (keyBytes, valueBytes) -> new KeyValue(
+                        Schema.STRING.decode(keyBytes),
+                        Schema.STRING.decode(valueBytes)
+                )
+        );
+        assertNotNull(kv);
```
Comment (sijie) [inline pulsar-client/src/test/java/org/apache/pulsar/common/schema/KeyValueTest.java:118]: Don't you need to assert the key & value of this `kv`?

**Review by sijie — CHANGES_REQUESTED** [review]

**Comment by eolivelli** [comment]
this patch is obsolete, we can close this PR
